From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Fri, 5 Dec 2025 15:22:17 +0100
Subject: [PATCH 082/225] iio: Add Unisoc sensor hub drivers
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The PMSYS coprocessor on Unisoc SoCs implements a sensor hub. Add a
common driver for communicating with the sensor hub and IIO drivers for
accelerometer, magnetometer, gyroscope, ambient light and proximity
sensors as well as a combined compass orientation sensor.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/iio/accel/Kconfig                    |   4 +
 drivers/iio/accel/Makefile                   |   2 +
 drivers/iio/accel/sprd_shub_accel.c          | 182 +++++
 drivers/iio/common/Kconfig                   |   1 +
 drivers/iio/common/Makefile                  |   1 +
 drivers/iio/common/sprd_shub/Kconfig         |  11 +
 drivers/iio/common/sprd_shub/Makefile        |   1 +
 drivers/iio/common/sprd_shub/sprd_shub.c     | 678 +++++++++++++++++++
 drivers/iio/common/sprd_shub/sprd_shub_msg.h |   6 +
 drivers/iio/gyro/Kconfig                     |   4 +
 drivers/iio/gyro/Makefile                    |   2 +
 drivers/iio/gyro/sprd_shub_gyro.c            | 182 +++++
 drivers/iio/light/Kconfig                    |   4 +
 drivers/iio/light/Makefile                   |   1 +
 drivers/iio/light/sprd_shub_light.c          | 142 ++++
 drivers/iio/magnetometer/Kconfig             |   8 +
 drivers/iio/magnetometer/Makefile            |   3 +
 drivers/iio/magnetometer/sprd_shub_compass.c | 158 +++++
 drivers/iio/magnetometer/sprd_shub_magn.c    | 166 +++++
 drivers/iio/proximity/Kconfig                |   4 +
 drivers/iio/proximity/Makefile               |   1 +
 drivers/iio/proximity/sprd_shub_proximity.c  | 140 ++++
 include/linux/iio/common/sprd_shub.h         |  38 ++
 23 files changed, 1739 insertions(+)
 create mode 100644 drivers/iio/accel/sprd_shub_accel.c
 create mode 100644 drivers/iio/common/sprd_shub/Kconfig
 create mode 100644 drivers/iio/common/sprd_shub/Makefile
 create mode 100644 drivers/iio/common/sprd_shub/sprd_shub.c
 create mode 100644 drivers/iio/common/sprd_shub/sprd_shub_msg.h
 create mode 100644 drivers/iio/gyro/sprd_shub_gyro.c
 create mode 100644 drivers/iio/light/sprd_shub_light.c
 create mode 100644 drivers/iio/magnetometer/sprd_shub_compass.c
 create mode 100644 drivers/iio/magnetometer/sprd_shub_magn.c
 create mode 100644 drivers/iio/proximity/sprd_shub_proximity.c
 create mode 100644 include/linux/iio/common/sprd_shub.h

diff --git a/drivers/iio/accel/Kconfig b/drivers/iio/accel/Kconfig
index 4094299e2..adae6209d 100644
--- a/drivers/iio/accel/Kconfig
+++ b/drivers/iio/accel/Kconfig
@@ -424,6 +424,10 @@ config IIO_CROS_EC_ACCEL_LEGACY
 	  Sensor data is retrieved through IO memory.
 	  Newer devices should use IIO_CROS_EC_SENSORS.
 
+config IIO_SPRD_SHUB_ACCEL
+	tristate "Unisoc sensor hub accelerometer driver"
+	depends on IIO_SPRD_SHUB
+
 config IIO_ST_ACCEL_3AXIS
 	tristate "STMicroelectronics accelerometers 3-Axis Driver"
 	depends on (I2C || SPI_MASTER) && SYSFS
diff --git a/drivers/iio/accel/Makefile b/drivers/iio/accel/Makefile
index fa440a859..0411277c1 100644
--- a/drivers/iio/accel/Makefile
+++ b/drivers/iio/accel/Makefile
@@ -80,6 +80,8 @@ obj-$(CONFIG_STK8BA50)		+= stk8ba50.o
 
 obj-$(CONFIG_IIO_CROS_EC_ACCEL_LEGACY) += cros_ec_accel_legacy.o
 
+obj-$(CONFIG_IIO_SPRD_SHUB_ACCEL) += sprd_shub_accel.o
+
 obj-$(CONFIG_IIO_SSP_SENSORS_COMMONS) += ssp_accel_sensor.o
 
 obj-$(CONFIG_IIO_ST_ACCEL_3AXIS) += st_accel.o
diff --git a/drivers/iio/accel/sprd_shub_accel.c b/drivers/iio/accel/sprd_shub_accel.c
new file mode 100644
index 000000000..c2f0e7b2f
--- /dev/null
+++ b/drivers/iio/accel/sprd_shub_accel.c
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Unisoc sensor hub IIO accelerometer driver
+ *
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/mod_devicetable.h>
+#include <linux/iio/common/sprd_shub.h>
+#include <linux/iio/kfifo_buf.h>
+#include <linux/platform_device.h>
+
+static const struct iio_chan_spec_ext_info shub_accel_ext_info[] = {
+	IIO_MOUNT_MATRIX(IIO_SHARED_BY_ALL, sprd_shub_get_mount_matrix),
+	{ }
+};
+
+static const struct iio_chan_spec shub_accel_iio_channels[] = {
+	{
+		.type = IIO_ACCEL,
+		.modified = true,
+		.channel2 = IIO_MOD_X,
+		.scan_index = 0,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_separate = BIT(IIO_CHAN_INFO_CALIBBIAS),
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+		.ext_info = shub_accel_ext_info,
+	},
+	{
+		.type = IIO_ACCEL,
+		.modified = true,
+		.channel2 = IIO_MOD_Y,
+		.scan_index = 1,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_separate = BIT(IIO_CHAN_INFO_CALIBBIAS),
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+		.ext_info = shub_accel_ext_info,
+	},
+	{
+		.type = IIO_ACCEL,
+		.modified = true,
+		.channel2 = IIO_MOD_Z,
+		.scan_index = 2,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_separate = BIT(IIO_CHAN_INFO_CALIBBIAS),
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+		.ext_info = shub_accel_ext_info,
+	},
+	{
+		.type = IIO_TIMESTAMP,
+		.channel = -1,
+		.scan_index = 3,
+		.scan_type = {
+			.sign = 'u',
+			.realbits = 64,
+			.storagebits = 64,
+			.endianness = IIO_CPU,
+		},
+	},
+};
+
+static int shub_accel_read_raw(struct iio_dev *iio_dev,
+			       const struct iio_chan_spec *chan, int *val,
+			       int *val2, long mask)
+{
+	int ret;
+
+	switch (mask) {
+	case IIO_CHAN_INFO_CALIBBIAS:
+		ret = sprd_shub_calib_read(iio_dev, chan->scan_index, val);
+		if (ret)
+			return ret;
+
+		*val2 = 10000;
+		return IIO_VAL_FRACTIONAL;
+
+	case IIO_CHAN_INFO_SCALE:
+		*val = 1;
+		*val2 = SPRD_SHUB_FIXED_SCALE;
+		return IIO_VAL_FRACTIONAL;
+
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_read_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static int shub_accel_write_raw(struct iio_dev *iio_dev,
+				const struct iio_chan_spec *chan, int val,
+				int val2, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_CALIBBIAS:
+		return sprd_shub_calib_write(iio_dev, chan->scan_index, val);
+
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_write_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static const struct iio_info shub_accel_iio_info = {
+	.read_raw = shub_accel_read_raw,
+	.write_raw = shub_accel_write_raw,
+	.read_avail = sprd_shub_read_avail,
+};
+
+static int sprd_shub_accel_probe(struct platform_device *pdev)
+{
+	struct iio_dev *iio_dev;
+	struct sprd_shub_sensor_priv *priv;
+	int ret;
+
+	iio_dev = devm_iio_device_alloc(&pdev->dev, sizeof(*priv));
+	if (!iio_dev)
+		return -ENOMEM;
+
+	priv = iio_priv(iio_dev);
+	priv->sensor = *(struct sprd_shub_sensor **)pdev->dev.platform_data;
+	priv->process = sprd_shub_process_xyz;
+
+	iio_dev->name = "sprd-shub-accel";
+	iio_dev->info = &shub_accel_iio_info;
+	iio_dev->channels = shub_accel_iio_channels;
+	iio_dev->num_channels = ARRAY_SIZE(shub_accel_iio_channels);
+
+	ret = devm_iio_kfifo_buffer_setup(&pdev->dev, iio_dev, &sprd_shub_buffer_ops);
+	if (ret)
+		return ret;
+
+	ret = devm_iio_device_register(&pdev->dev, iio_dev);
+	if (ret)
+		return ret;
+
+	platform_set_drvdata(pdev, iio_dev);
+
+	return 0;
+}
+
+static void sprd_shub_accel_remove(struct platform_device *pdev)
+{
+	platform_set_drvdata(pdev, NULL);
+}
+
+static const struct platform_device_id sprd_shub_accel_ids[] = {
+	{ .name = "sprd-shub-accel" },
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, sprd_shub_accel_ids);
+
+static struct platform_driver sprd_shub_accel_driver = {
+	.probe = sprd_shub_accel_probe,
+	.remove = sprd_shub_accel_remove,
+	.driver	= {
+		.name = "sprd_shub_accel",
+	},
+	.id_table = sprd_shub_accel_ids,
+};
+module_platform_driver(sprd_shub_accel_driver);
+
+MODULE_DESCRIPTION("Unisoc sensor hub accelerometer driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/iio/common/Kconfig b/drivers/iio/common/Kconfig
index 1ccb5ccf3..856ae28cd 100644
--- a/drivers/iio/common/Kconfig
+++ b/drivers/iio/common/Kconfig
@@ -8,5 +8,6 @@ source "drivers/iio/common/hid-sensors/Kconfig"
 source "drivers/iio/common/inv_sensors/Kconfig"
 source "drivers/iio/common/ms_sensors/Kconfig"
 source "drivers/iio/common/scmi_sensors/Kconfig"
+source "drivers/iio/common/sprd_shub/Kconfig"
 source "drivers/iio/common/ssp_sensors/Kconfig"
 source "drivers/iio/common/st_sensors/Kconfig"
diff --git a/drivers/iio/common/Makefile b/drivers/iio/common/Makefile
index d3e952239..06a149486 100644
--- a/drivers/iio/common/Makefile
+++ b/drivers/iio/common/Makefile
@@ -13,5 +13,6 @@ obj-y += hid-sensors/
 obj-y += inv_sensors/
 obj-y += ms_sensors/
 obj-y += scmi_sensors/
+obj-y += sprd_shub/
 obj-y += ssp_sensors/
 obj-y += st_sensors/
diff --git a/drivers/iio/common/sprd_shub/Kconfig b/drivers/iio/common/sprd_shub/Kconfig
new file mode 100644
index 000000000..c3935d7f6
--- /dev/null
+++ b/drivers/iio/common/sprd_shub/Kconfig
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: GPL-2.0-only
+
+config IIO_SPRD_SHUB
+	tristate "Unisoc sensor hub support"
+	depends on ARCH_SPRD || COMPILE_TEST
+	depends on RPMSG
+	select IIO_BUFFER
+	select IIO_KFIFO_BUFFER
+	help
+	  This module exposes the sensor hub on the PMSYS coprocessor found
+	  in Unisoc SoCs as an IIO device.
diff --git a/drivers/iio/common/sprd_shub/Makefile b/drivers/iio/common/sprd_shub/Makefile
new file mode 100644
index 000000000..70492e7d3
--- /dev/null
+++ b/drivers/iio/common/sprd_shub/Makefile
@@ -0,0 +1 @@
+obj-$(CONFIG_IIO_SPRD_SHUB) += sprd_shub.o
diff --git a/drivers/iio/common/sprd_shub/sprd_shub.c b/drivers/iio/common/sprd_shub/sprd_shub.c
new file mode 100644
index 000000000..1c6ec7025
--- /dev/null
+++ b/drivers/iio/common/sprd_shub/sprd_shub.c
@@ -0,0 +1,678 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Unisoc sensor hub IIO driver
+ *
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/iio/buffer.h>
+#include <linux/iio/common/sprd_shub.h>
+#include <linux/platform_device.h>
+#include <linux/rpmsg.h>
+#include <linux/unaligned.h>
+
+#define SHUB_MSG_MAGIC_CHAR		'~'
+#define SHUB_MSG_MAGIC_COUNT		4
+
+enum {
+	SHUB_ACCELEROMETER_SENSOR = 1,
+	SHUB_GEOMAGNETIC_SENSOR = 2,
+	SHUB_ORIENTATION_SENSOR = 3,
+	SHUB_GYROSCOPE_SENSOR = 4,
+	SHUB_LIGHT_SENSOR = 5,
+	SHUB_PROXIMITY_SENSOR = 8,
+};
+
+struct shub_msg_hdr {
+	u8 magic[SHUB_MSG_MAGIC_COUNT];
+	u8 sensor;
+	u8 type;
+	__be16 length;
+	__be16 checksum;
+} __packed;
+
+struct shub_batch_cmd {
+	__le32 sample_interval_us;
+	__le32 batch_interval_us;
+	__le32 reserved;
+} __packed;
+
+struct shub_data_evt {
+	u8 handle;
+	u8 reserved;
+	u8 data[24];
+	__le64 timestamp;
+} __packed;
+
+struct shub_sensor_info {
+	char model[20];
+	char manufacturer[20];
+	__le32 version;
+	__le32 handle;
+	__le32 maxrange;
+	__le32 resolution;
+	__le32 power;
+	__le32 mindelay_us;
+	__le32 fifo_reserved_event_count;
+	__le32 fifo_max_event_count;
+	__le32 maxdelay_us;
+} __packed;
+
+struct shub_timesync_cmd {
+	__le64 boottime;
+	__le64 systimer_cnt;
+	__le64 sysfrt_cnt;
+	__le32 systimer_mult;
+	__le32 systimer_shift;
+	__le32 sysfrt_mult;
+	__le32 sysfrt_shift;
+} __packed;
+
+enum {
+	SHUB_CMD_ENABLE = 0x00,
+	SHUB_CMD_DISABLE = 0x01,
+	SHUB_CMD_SET_BATCH = 0x02,
+	SHUB_EVT_DATA = 0x04,
+	SHUB_CMD_GET_SENSOR_INFO = 0x09,
+	SHUB_CMD_SET_CALIB_DATA = 0xc2,
+	SHUB_CMD_START_CALIBRATION = 0xc3,
+	SHUB_CMD_SET_TIMESYNC = 0xc4,
+	SHUB_CMD_GET_CALIB_DATA = 0xc6,
+	SHUB_CMD_GET_FWVERSION = 0xcd,
+};
+
+enum shub_msg_rx_state {
+	SHUB_MSG_RX_SYNC,
+	SHUB_MSG_RX_HEADER,
+	SHUB_MSG_RX_DATA,
+};
+
+struct sprd_shub {
+	struct device *dev;
+	struct rpmsg_endpoint *ept;
+
+	enum shub_msg_rx_state rx_state;
+	size_t rx_pos;
+	struct shub_msg_hdr rx_hdr;
+	u8 rx_data[256];
+
+	struct delayed_work init_work;
+	spinlock_t lock;
+	struct idr sensors;
+};
+
+struct sprd_shub_sensor {
+	struct sprd_shub *shub;
+	struct platform_device *pdev;
+	struct work_struct reg_work;
+	u8 id;
+};
+
+static u16 shub_msg_checksum(const void *data, size_t len)
+{
+	const __be16 *buf = data;
+	u32 sum = 0;
+
+	while (len >= 2) {
+		sum += be16_to_cpu(*buf++);
+		len -= 2;
+	}
+
+	if (len == 1)
+		sum += be16_to_cpu(*buf) & 0xff00;
+
+	sum = (sum & 0xffff) + (sum >> 16);
+	sum = (sum & 0xffff) + (sum >> 16);
+
+	return ~sum;
+}
+
+static const char *shub_sensor_type_names[] = {
+	[SHUB_ACCELEROMETER_SENSOR] = "sprd-shub-accel",
+	[SHUB_GEOMAGNETIC_SENSOR] = "sprd-shub-magn",
+	[SHUB_ORIENTATION_SENSOR] = "sprd-shub-compass",
+	[SHUB_GYROSCOPE_SENSOR] = "sprd-shub-gyro",
+	[SHUB_LIGHT_SENSOR] = "sprd-shub-light",
+	[SHUB_PROXIMITY_SENSOR] = "sprd-shub-proximity",
+};
+
+static void shub_sensor_register(struct work_struct *work)
+{
+	struct sprd_shub_sensor *sensor =
+		container_of(work, struct sprd_shub_sensor, reg_work);
+	struct sprd_shub *shub = sensor->shub;
+	struct platform_device *pdev;
+	const char *name = NULL;
+
+	if (sensor->id < ARRAY_SIZE(shub_sensor_type_names))
+		name = shub_sensor_type_names[sensor->id];
+
+	if (!name) {
+		dev_warn(shub->dev, "unknown sensor ID %d\n", sensor->id);
+		return;
+	}
+
+	pdev = platform_device_register_data(shub->dev, name, sensor->id,
+					     &sensor, sizeof(sensor));
+	if (IS_ERR(pdev))
+		dev_err(shub->dev, "failed to register %s: %pe\n", name, pdev);
+
+	spin_lock_irq(&shub->lock);
+	sensor->pdev = pdev;
+	spin_unlock_irq(&shub->lock);
+}
+
+static void shub_add_sensor(struct sprd_shub *shub, u8 sensor_id)
+{
+	struct sprd_shub_sensor *sensor;
+	unsigned long flags;
+	int ret;
+
+	sensor = kzalloc(sizeof(*sensor), GFP_ATOMIC);
+	if (!sensor)
+		return;
+
+	sensor->shub = shub;
+	sensor->id = sensor_id;
+	INIT_WORK(&sensor->reg_work, shub_sensor_register);
+
+	spin_lock_irqsave(&shub->lock, flags);
+	ret = idr_alloc(&shub->sensors, sensor, sensor_id, sensor_id + 1, GFP_ATOMIC);
+	if (ret == sensor_id) {
+		schedule_work(&sensor->reg_work);
+	} else {
+		dev_err(shub->dev, "sensor %d already allocated\n", sensor_id);
+		kfree(sensor);
+	}
+	spin_unlock_irqrestore(&shub->lock, flags);
+}
+
+static void shub_handle_sensor_info(struct sprd_shub *shub)
+{
+	struct shub_sensor_info *info = (void *)shub->rx_data;
+	u8 sensor_id = le32_to_cpu(info->handle);
+
+	shub_add_sensor(shub, sensor_id);
+}
+
+static void shub_handle_data(struct sprd_shub *shub)
+{
+	struct shub_data_evt *evt = (void *)shub->rx_data;
+	struct sprd_shub_sensor_priv *priv;
+	struct sprd_shub_sensor *sensor;
+	struct iio_dev *iio_dev;
+	unsigned long flags;
+
+	spin_lock_irqsave(&shub->lock, flags);
+
+	sensor = idr_find(&shub->sensors, evt->handle - 100);
+	if (!sensor || !sensor->pdev)
+		goto out;
+
+	iio_dev = platform_get_drvdata(sensor->pdev);
+	if (!iio_dev)
+		goto out;
+
+	priv = iio_priv(iio_dev);
+	priv->process(iio_dev, evt->data, le64_to_cpu(evt->timestamp));
+
+out:
+	spin_unlock_irqrestore(&shub->lock, flags);
+}
+
+static void shub_handle_msg(struct sprd_shub *shub)
+{
+	dev_dbg(shub->dev, "event: 0x%02x 0x%02x, data size 0x%04x\n",
+		shub->rx_hdr.sensor, shub->rx_hdr.type,
+		be16_to_cpu(shub->rx_hdr.length));
+
+	switch (shub->rx_hdr.type) {
+	case SHUB_EVT_DATA:
+		shub_handle_data(shub);
+		break;
+
+	case SHUB_CMD_GET_SENSOR_INFO:
+		shub_handle_sensor_info(shub);
+		break;
+
+	case SHUB_CMD_GET_FWVERSION:
+		/* Give the hub some time to initialize */
+		schedule_delayed_work(&shub->init_work, 2 * HZ);
+		break;
+	}
+}
+
+static size_t shub_rx_buf(u8 **buf, int *cnt, void *to, size_t *pos,
+			  size_t expected, size_t avail)
+{
+	size_t len = min_t(size_t, *cnt, expected - *pos);
+
+	if (*pos + len <= avail)
+		memcpy(to + *pos, *buf, len);
+	else if (*pos < avail)
+		memcpy(to + *pos, *buf, avail - *pos);
+
+	*pos += len;
+	*buf += len;
+	*cnt -= len;
+
+	return expected - *pos;
+}
+
+static int shub_rpmsg_callback(struct rpmsg_device *rpdev, void *data,
+			       int cnt, void *userdata, u32 addr)
+{
+	struct sprd_shub *shub = userdata;
+	u8 *buf = data;
+
+	while (cnt > 0) {
+		switch (shub->rx_state) {
+		case SHUB_MSG_RX_SYNC:
+			while (cnt && shub->rx_pos < SHUB_MSG_MAGIC_COUNT) {
+				if (*buf == SHUB_MSG_MAGIC_CHAR)
+					shub->rx_pos++;
+				else
+					shub->rx_pos = 0;
+				buf++;
+				cnt--;
+			}
+
+			if (cnt == 0)
+				break;
+
+			shub->rx_state = SHUB_MSG_RX_HEADER;
+			fallthrough;
+
+		case SHUB_MSG_RX_HEADER:
+			if (shub_rx_buf(&buf, &cnt, &shub->rx_hdr, &shub->rx_pos,
+					sizeof(struct shub_msg_hdr),
+					sizeof(struct shub_msg_hdr)))
+				return 0;
+
+			shub->rx_pos = 0;
+			shub->rx_state = SHUB_MSG_RX_DATA;
+			break;
+
+		case SHUB_MSG_RX_DATA:
+			if (shub_rx_buf(&buf, &cnt, shub->rx_data, &shub->rx_pos,
+					be16_to_cpu(shub->rx_hdr.length),
+					sizeof(shub->rx_data)))
+				return 0;
+
+			shub_handle_msg(shub);
+
+			shub->rx_pos = 0;
+			shub->rx_state = SHUB_MSG_RX_SYNC;
+			break;
+		}
+	}
+
+	return 0;
+}
+
+static int shub_send_msg(struct sprd_shub *shub, u8 sensor, u8 type,
+			 void *data, size_t data_size)
+{
+	static u8 dummy_data = { 0xff };
+	struct shub_msg_hdr hdr;
+	__be16 checksum;
+	int ret;
+
+	if (data_size > U16_MAX)
+		return -EINVAL;
+
+	if (data_size == 0) {
+		data = &dummy_data;
+		data_size = 1;
+	}
+
+	dev_dbg(shub->dev, "send command: 0x%02x 0x%02x, data size 0x%04zx\n",
+		sensor, type, data_size);
+
+	memset(hdr.magic, SHUB_MSG_MAGIC_CHAR, SHUB_MSG_MAGIC_COUNT);
+	hdr.sensor = sensor;
+	hdr.type = type;
+	hdr.length = cpu_to_be16(data_size);
+	hdr.checksum = cpu_to_be16(shub_msg_checksum(&hdr, sizeof(hdr) - 2));
+
+	ret = rpmsg_send(shub->ept, &hdr, sizeof(hdr));
+	if (ret) {
+		dev_err(shub->dev, "failed to send command header: %d\n", ret);
+		return ret;
+	}
+
+	ret = rpmsg_send(shub->ept, data, data_size);
+	if (ret) {
+		dev_err(shub->dev, "failed to send data: %d\n", ret);
+		return ret;
+	}
+
+	checksum = cpu_to_be16(shub_msg_checksum(data, data_size));
+
+	ret = rpmsg_send(shub->ept, &checksum, sizeof(checksum));
+	if (ret) {
+		dev_err(shub->dev, "failed to send data: %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static const int sprd_shub_frequencies[][2] = {
+	{   3, 125000 },
+	{   6, 250000 },
+	{  12, 500000 },
+	{  25, 0 },
+	{  50, 0 },
+	{ 100, 0 },
+	{ 200, 0 },
+	{ 500, 0 },
+};
+
+static const int sprd_shub_intervals[] = {
+	320000,
+	160000,
+	80000,
+	40000,
+	20000,
+	10000,
+	5000,
+	2000,
+};
+
+static int shub_sensor_postenable(struct iio_dev *iio_dev)
+{
+	struct sprd_shub_sensor_priv *priv = iio_priv(iio_dev);
+	struct sprd_shub_sensor *sensor = priv->sensor;
+	struct sprd_shub *shub = sensor->shub;
+	struct shub_batch_cmd batch_cmd = {
+		.sample_interval_us = sprd_shub_intervals[priv->freq_idx],
+		.batch_interval_us = 0,
+	};
+	int ret;
+
+	spin_lock_irq(&shub->lock);
+	priv->scan_mask = *iio_dev->active_scan_mask;
+	spin_unlock_irq(&shub->lock);
+
+	ret = shub_send_msg(shub, 100 + sensor->id, SHUB_CMD_SET_BATCH,
+			    &batch_cmd, sizeof(batch_cmd));
+	if (ret)
+		return ret;
+
+	return shub_send_msg(shub, 100 + sensor->id, SHUB_CMD_ENABLE, NULL, 0);
+}
+
+static int shub_sensor_postdisable(struct iio_dev *iio_dev)
+{
+	struct sprd_shub_sensor_priv *priv = iio_priv(iio_dev);
+	struct sprd_shub_sensor *sensor = priv->sensor;
+	struct sprd_shub *shub = sensor->shub;
+
+	return shub_send_msg(shub, 100 + sensor->id, SHUB_CMD_DISABLE, NULL, 0);
+}
+
+const struct iio_buffer_setup_ops sprd_shub_buffer_ops = {
+	.postenable = &shub_sensor_postenable,
+	.postdisable = &shub_sensor_postdisable
+};
+EXPORT_SYMBOL_GPL(sprd_shub_buffer_ops);
+
+static const struct iio_mount_matrix shub_default_mount_matrix = {
+	.rotation = {
+		"-1", "0", "0",
+		"0", "-1", "0",
+		"0", "0", "1",
+	},
+};
+
+const struct iio_mount_matrix *
+sprd_shub_get_mount_matrix(const struct iio_dev *iio_dev,
+			   const struct iio_chan_spec *chan)
+{
+	return &shub_default_mount_matrix;
+}
+EXPORT_SYMBOL_GPL(sprd_shub_get_mount_matrix);
+
+int sprd_shub_calib_read(struct iio_dev *iio_dev, int idx, int *val)
+{
+	struct sprd_shub_sensor_priv *priv = iio_priv(iio_dev);
+
+	if (idx >= ARRAY_SIZE(priv->calib_data))
+		return -EINVAL;
+
+	*val = priv->calib_data[idx];
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(sprd_shub_calib_read);
+
+int sprd_shub_calib_write(struct iio_dev *iio_dev, int idx, int val)
+{
+	struct sprd_shub_sensor_priv *priv = iio_priv(iio_dev);
+	struct sprd_shub_sensor *sensor = priv->sensor;
+	struct sprd_shub *shub = sensor->shub;
+
+	if (idx >= ARRAY_SIZE(priv->calib_data))
+		return -EINVAL;
+
+	priv->calib_data[idx] = val;
+
+	return shub_send_msg(shub, 100 + sensor->id, SHUB_CMD_SET_CALIB_DATA,
+			     priv->calib_data, sizeof(priv->calib_data));
+}
+EXPORT_SYMBOL_GPL(sprd_shub_calib_write);
+
+int sprd_shub_read_freq(struct iio_dev *iio_dev, int *val, int *val2)
+{
+	struct sprd_shub_sensor_priv *priv = iio_priv(iio_dev);
+
+	*val = sprd_shub_frequencies[priv->freq_idx][0];
+	*val2 = sprd_shub_frequencies[priv->freq_idx][1];
+
+	return IIO_VAL_INT_PLUS_MICRO;
+}
+EXPORT_SYMBOL_GPL(sprd_shub_read_freq);
+
+int sprd_shub_write_freq(struct iio_dev *iio_dev, int val, int val2)
+{
+	struct sprd_shub_sensor_priv *priv = iio_priv(iio_dev);
+	struct sprd_shub_sensor *sensor = priv->sensor;
+	struct sprd_shub *shub = sensor->shub;
+	struct shub_batch_cmd batch_cmd = {
+		.batch_interval_us = 0,
+	};
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(sprd_shub_frequencies); i++) {
+		if (sprd_shub_frequencies[i][0] == val &&
+		    sprd_shub_frequencies[i][1] == val2) {
+			priv->freq_idx = i;
+			batch_cmd.sample_interval_us = sprd_shub_intervals[i];
+			break;
+		}
+	}
+
+	if (i == ARRAY_SIZE(sprd_shub_frequencies))
+		return -EINVAL;
+
+	return shub_send_msg(shub, 100 + sensor->id, SHUB_CMD_SET_BATCH,
+			     &batch_cmd, sizeof(batch_cmd));
+}
+EXPORT_SYMBOL_GPL(sprd_shub_write_freq);
+
+int sprd_shub_read_avail(struct iio_dev *iio_dev,
+			 struct iio_chan_spec const *chan,
+			 const int **vals, int *type,
+			 int *length, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		*length = ARRAY_SIZE(sprd_shub_frequencies) * 2;
+		*vals = &sprd_shub_frequencies[0][0];
+		*type = IIO_VAL_INT_PLUS_MICRO;
+		return IIO_AVAIL_LIST;
+	}
+
+	return -EINVAL;
+}
+EXPORT_SYMBOL_GPL(sprd_shub_read_avail);
+
+static s32 shub_float_to_int(u32 float32)
+{
+	int fraction, shift,
+	    scale = SPRD_SHUB_FIXED_SCALE,
+	    mantissa = float32 & GENMASK(22, 0),
+	    sign = (float32 & BIT(31)) ? -1 : 1,
+	    exp = (float32 & ~BIT(31)) >> 23;
+
+	/* special case 0 */
+	if (!exp && !mantissa)
+		return 0;
+
+	exp -= 127;
+	if (exp < 0) {
+		exp = -exp;
+		/* return values ranging from 1 to scale-1 */
+		return sign * ((((BIT(23) + mantissa) * scale) >> 23) >> exp);
+	}
+
+	/* return values starting at 1*scale */
+	shift = 23 - exp;
+	float32 = BIT(exp) + (mantissa >> shift);
+	fraction = mantissa & GENMASK(shift - 1, 0);
+
+	return sign * (float32 * scale + ((fraction * scale) >> shift));
+}
+
+void sprd_shub_process_xyz(struct iio_dev *iio_dev, void *data, u64 timestamp)
+{
+	struct sprd_shub_sensor_priv *priv = iio_priv(iio_dev);
+	s32 buf[6] __aligned(8);
+	s32 *p = buf;
+	__le32 *flt = data;
+	unsigned int i;
+
+	memset(buf, 0, sizeof(buf));
+
+	for_each_set_bit(i, &priv->scan_mask, iio_get_masklength(iio_dev)) {
+		if (i < 3)
+			*p++ = shub_float_to_int(get_unaligned_le32(&flt[i]));
+	}
+
+	iio_push_to_buffers_with_timestamp(iio_dev, buf, timestamp);
+}
+EXPORT_SYMBOL_GPL(sprd_shub_process_xyz);
+
+void sprd_shub_process_single(struct iio_dev *iio_dev, void *data, u64 timestamp)
+{
+	struct sprd_shub_sensor_priv *priv = iio_priv(iio_dev);
+	s32 buf[4] __aligned(8);
+
+	memset(buf, 0, sizeof(buf));
+
+	if (test_bit(0, &priv->scan_mask))
+		buf[0] = shub_float_to_int(get_unaligned_le32(data));
+
+	iio_push_to_buffers_with_timestamp(iio_dev, buf, timestamp);
+}
+EXPORT_SYMBOL_GPL(sprd_shub_process_single);
+
+static void shub_initialize(struct work_struct *work)
+{
+	struct sprd_shub *shub = container_of(to_delayed_work(work),
+					      struct sprd_shub, init_work);
+
+	/*
+	 * This command makes the sensor hub send an info event for
+	 * each sensor, which we handle by registering the sensor.
+	 */
+	shub_send_msg(shub, 0xff, SHUB_CMD_GET_SENSOR_INFO, NULL, 0);
+}
+
+static int shub_rpmsg_probe(struct rpmsg_device *rpdev)
+{
+	struct rpmsg_channel_info chinfo = {
+		.src = RPMSG_ADDR_ANY,
+		.dst = RPMSG_ADDR_ANY,
+	};
+	struct shub_timesync_cmd timesync = {
+		.systimer_mult = cpu_to_le32(4096000000),
+		.systimer_shift = cpu_to_le32(12),
+		.sysfrt_mult = cpu_to_le32(4000000000),
+		.sysfrt_shift = cpu_to_le32(17),
+	};
+	struct sprd_shub *shub;
+	int ret;
+
+	shub = devm_kzalloc(&rpdev->dev, sizeof(*shub), GFP_KERNEL);
+	if (!shub)
+		return -ENOMEM;
+
+	shub->dev = &rpdev->dev;
+	dev_set_drvdata(&rpdev->dev, shub);
+
+	spin_lock_init(&shub->lock);
+	idr_init(&shub->sensors);
+	INIT_DELAYED_WORK(&shub->init_work, shub_initialize);
+
+	strscpy(chinfo.name, "shub_main", sizeof(chinfo.name));
+	shub->ept = rpmsg_create_ept(rpdev, shub_rpmsg_callback, shub, chinfo);
+	if (!shub->ept)
+		return -ENOMEM;
+
+	ret = shub_send_msg(shub, 0xff, SHUB_CMD_SET_TIMESYNC,
+			    &timesync, sizeof(timesync));
+	if (ret)
+		return ret;
+
+	ret = shub_send_msg(shub, 0xff, SHUB_CMD_GET_FWVERSION, NULL, 0);
+	if (ret)
+		return ret;
+
+	/*
+	 * Register virtual sensors.
+	 * FIXME: These sensors may depend on multiple physical sensors.
+	 * If these are not present on some devices, the corresponding
+	 * virtual sensors should not be registered.
+	 */
+	shub_add_sensor(shub, SHUB_ORIENTATION_SENSOR);
+
+	return 0;
+}
+
+static void shub_rpmsg_remove(struct rpmsg_device *rpdev)
+{
+	struct sprd_shub *shub = dev_get_drvdata(&rpdev->dev);
+	struct sprd_shub_sensor *sensor;
+	int sensor_id;
+
+	rpmsg_destroy_ept(shub->ept);
+
+	idr_for_each_entry(&shub->sensors, sensor, sensor_id) {
+		cancel_work_sync(&sensor->reg_work);
+		platform_device_unregister(sensor->pdev);
+		kfree(sensor);
+	}
+	idr_destroy(&shub->sensors);
+}
+
+static const struct of_device_id shub_of_match[] = {
+	{ .compatible = "sprd,shub" },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, shub_of_match);
+
+static struct rpmsg_driver shub_rpmsg_driver = {
+	.probe = shub_rpmsg_probe,
+	.remove = shub_rpmsg_remove,
+	.drv = {
+		.name = "sprd-shub",
+		.of_match_table = shub_of_match,
+	},
+};
+
+module_rpmsg_driver(shub_rpmsg_driver);
+
+MODULE_DESCRIPTION("Unisoc sensor hub IIO driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/iio/common/sprd_shub/sprd_shub_msg.h b/drivers/iio/common/sprd_shub/sprd_shub_msg.h
new file mode 100644
index 000000000..db68e39cd
--- /dev/null
+++ b/drivers/iio/common/sprd_shub/sprd_shub_msg.h
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Unisoc sensor hub driver
+ *
+ * Copyright (C) 2025 Otto Pflüger
+ */
diff --git a/drivers/iio/gyro/Kconfig b/drivers/iio/gyro/Kconfig
index 3e193ee0f..ed0a93116 100644
--- a/drivers/iio/gyro/Kconfig
+++ b/drivers/iio/gyro/Kconfig
@@ -135,6 +135,10 @@ config MPU3050_I2C
 	  This driver can be built as a module. The module will be called
 	  inv-mpu3050-i2c.
 
+config IIO_SPRD_SHUB_GYRO
+	tristate "Unisoc sensor hub gyroscope driver"
+	depends on IIO_SPRD_SHUB
+
 config IIO_ST_GYRO_3AXIS
 	tristate "STMicroelectronics gyroscopes 3-Axis Driver"
 	depends on (I2C || SPI_MASTER) && SYSFS
diff --git a/drivers/iio/gyro/Makefile b/drivers/iio/gyro/Makefile
index 0319b397d..3d46a0edf 100644
--- a/drivers/iio/gyro/Makefile
+++ b/drivers/iio/gyro/Makefile
@@ -28,6 +28,8 @@ itg3200-y               := itg3200_core.o
 itg3200-$(CONFIG_IIO_BUFFER) += itg3200_buffer.o
 obj-$(CONFIG_ITG3200)   += itg3200.o
 
+obj-$(CONFIG_IIO_SPRD_SHUB_GYRO) += sprd_shub_gyro.o
+
 obj-$(CONFIG_IIO_SSP_SENSORS_COMMONS) += ssp_gyro_sensor.o
 
 obj-$(CONFIG_IIO_ST_GYRO_3AXIS) += st_gyro.o
diff --git a/drivers/iio/gyro/sprd_shub_gyro.c b/drivers/iio/gyro/sprd_shub_gyro.c
new file mode 100644
index 000000000..3a92b3dcc
--- /dev/null
+++ b/drivers/iio/gyro/sprd_shub_gyro.c
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Unisoc sensor hub IIO gyroscope driver
+ *
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/mod_devicetable.h>
+#include <linux/iio/common/sprd_shub.h>
+#include <linux/iio/kfifo_buf.h>
+#include <linux/platform_device.h>
+
+static const struct iio_chan_spec_ext_info shub_gyro_ext_info[] = {
+	IIO_MOUNT_MATRIX(IIO_SHARED_BY_ALL, sprd_shub_get_mount_matrix),
+	{ }
+};
+
+static const struct iio_chan_spec shub_gyro_iio_channels[] = {
+	{
+		.type = IIO_ANGL_VEL,
+		.modified = true,
+		.channel2 = IIO_MOD_X,
+		.scan_index = 0,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_separate = BIT(IIO_CHAN_INFO_CALIBBIAS),
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+		.ext_info = shub_gyro_ext_info,
+	},
+	{
+		.type = IIO_ANGL_VEL,
+		.modified = true,
+		.channel2 = IIO_MOD_Y,
+		.scan_index = 1,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_separate = BIT(IIO_CHAN_INFO_CALIBBIAS),
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+		.ext_info = shub_gyro_ext_info,
+	},
+	{
+		.type = IIO_ANGL_VEL,
+		.modified = true,
+		.channel2 = IIO_MOD_Z,
+		.scan_index = 2,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_separate = BIT(IIO_CHAN_INFO_CALIBBIAS),
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+		.ext_info = shub_gyro_ext_info,
+	},
+	{
+		.type = IIO_TIMESTAMP,
+		.channel = -1,
+		.scan_index = 3,
+		.scan_type = {
+			.sign = 'u',
+			.realbits = 64,
+			.storagebits = 64,
+			.endianness = IIO_CPU,
+		},
+	},
+};
+
+static int shub_gyro_read_raw(struct iio_dev *iio_dev,
+			      const struct iio_chan_spec *chan, int *val,
+			      int *val2, long mask)
+{
+	int ret;
+
+	switch (mask) {
+	case IIO_CHAN_INFO_CALIBBIAS:
+		ret = sprd_shub_calib_read(iio_dev, chan->scan_index, val);
+		if (ret)
+			return ret;
+
+		*val2 = 10000;
+		return IIO_VAL_FRACTIONAL;
+
+	case IIO_CHAN_INFO_SCALE:
+		*val = 1;
+		*val2 = SPRD_SHUB_FIXED_SCALE;
+		return IIO_VAL_FRACTIONAL;
+
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_read_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static int shub_gyro_write_raw(struct iio_dev *iio_dev,
+			       const struct iio_chan_spec *chan, int val,
+			       int val2, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_CALIBBIAS:
+		return sprd_shub_calib_write(iio_dev, chan->scan_index, val);
+
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_write_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static const struct iio_info shub_gyro_iio_info = {
+	.read_raw = shub_gyro_read_raw,
+	.write_raw = shub_gyro_write_raw,
+	.read_avail = sprd_shub_read_avail,
+};
+
+static int sprd_shub_gyro_probe(struct platform_device *pdev)
+{
+	struct iio_dev *iio_dev;
+	struct sprd_shub_sensor_priv *priv;
+	int ret;
+
+	iio_dev = devm_iio_device_alloc(&pdev->dev, sizeof(*priv));
+	if (!iio_dev)
+		return -ENOMEM;
+
+	priv = iio_priv(iio_dev);
+	priv->sensor = *(struct sprd_shub_sensor **)pdev->dev.platform_data;
+	priv->process = sprd_shub_process_xyz;
+
+	iio_dev->name = "sprd-shub-gyro";
+	iio_dev->info = &shub_gyro_iio_info;
+	iio_dev->channels = shub_gyro_iio_channels;
+	iio_dev->num_channels = ARRAY_SIZE(shub_gyro_iio_channels);
+
+	ret = devm_iio_kfifo_buffer_setup(&pdev->dev, iio_dev, &sprd_shub_buffer_ops);
+	if (ret)
+		return ret;
+
+	ret = devm_iio_device_register(&pdev->dev, iio_dev);
+	if (ret)
+		return ret;
+
+	platform_set_drvdata(pdev, iio_dev);
+
+	return 0;
+}
+
+static void sprd_shub_gyro_remove(struct platform_device *pdev)
+{
+	platform_set_drvdata(pdev, NULL);
+}
+
+static const struct platform_device_id sprd_shub_gyro_ids[] = {
+	{ .name = "sprd-shub-gyro" },
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, sprd_shub_gyro_ids);
+
+static struct platform_driver sprd_shub_gyro_driver = {
+	.probe = sprd_shub_gyro_probe,
+	.remove = sprd_shub_gyro_remove,
+	.driver	= {
+		.name = "sprd_shub_gyro",
+	},
+	.id_table = sprd_shub_gyro_ids,
+};
+module_platform_driver(sprd_shub_gyro_driver);
+
+MODULE_DESCRIPTION("Unisoc sensor hub gyroscope driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/iio/light/Kconfig b/drivers/iio/light/Kconfig
index eff33e456..c8b54f69f 100644
--- a/drivers/iio/light/Kconfig
+++ b/drivers/iio/light/Kconfig
@@ -230,6 +230,10 @@ config IIO_CROS_EC_LIGHT_PROX
 	  To compile this driver as a module, choose M here:
 	  the module will be called cros_ec_light_prox.
 
+config IIO_SPRD_SHUB_LIGHT
+	tristate "Unisoc sensor hub ambient light sensor driver"
+	depends on IIO_SPRD_SHUB
+
 config GP2AP002
 	tristate "Sharp GP2AP002 Proximity/ALS sensor"
 	depends on I2C
diff --git a/drivers/iio/light/Makefile b/drivers/iio/light/Makefile
index c0048e0d5..f81ecf0b4 100644
--- a/drivers/iio/light/Makefile
+++ b/drivers/iio/light/Makefile
@@ -50,6 +50,7 @@ obj-$(CONFIG_ROHM_BU27034)	+= rohm-bu27034.o
 obj-$(CONFIG_RPR0521)		+= rpr0521.o
 obj-$(CONFIG_SI1133)		+= si1133.o
 obj-$(CONFIG_SI1145)		+= si1145.o
+obj-$(CONFIG_IIO_SPRD_SHUB_LIGHT) += sprd_shub_light.o
 obj-$(CONFIG_STK3310)          += stk3310.o
 obj-$(CONFIG_ST_UVIS25)		+= st_uvis25_core.o
 obj-$(CONFIG_ST_UVIS25_I2C)	+= st_uvis25_i2c.o
diff --git a/drivers/iio/light/sprd_shub_light.c b/drivers/iio/light/sprd_shub_light.c
new file mode 100644
index 000000000..8d9193cf9
--- /dev/null
+++ b/drivers/iio/light/sprd_shub_light.c
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Unisoc sensor hub IIO ambient light sensor driver
+ *
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/mod_devicetable.h>
+#include <linux/iio/common/sprd_shub.h>
+#include <linux/iio/kfifo_buf.h>
+#include <linux/platform_device.h>
+
+static const struct iio_chan_spec shub_light_iio_channels[] = {
+	{
+		.type = IIO_LIGHT,
+		.scan_index = 0,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_separate = BIT(IIO_CHAN_INFO_SCALE) |
+				      BIT(IIO_CHAN_INFO_CALIBSCALE) |
+				      BIT(IIO_CHAN_INFO_SAMP_FREQ),
+	},
+	{
+		.type = IIO_TIMESTAMP,
+		.channel = -1,
+		.scan_index = 1,
+		.scan_type = {
+			.sign = 'u',
+			.realbits = 64,
+			.storagebits = 64,
+			.endianness = IIO_CPU,
+		},
+	},
+};
+
+static int shub_light_read_raw(struct iio_dev *iio_dev,
+			       const struct iio_chan_spec *chan, int *val,
+			       int *val2, long mask)
+{
+	int ret;
+
+	switch (mask) {
+	case IIO_CHAN_INFO_CALIBSCALE:
+		ret = sprd_shub_calib_read(iio_dev, 0, val);
+		if (ret)
+			return ret;
+
+		*val2 = 10000;
+		return IIO_VAL_FRACTIONAL;
+
+	case IIO_CHAN_INFO_SCALE:
+		*val = 1;
+		*val2 = SPRD_SHUB_FIXED_SCALE;
+		return IIO_VAL_FRACTIONAL;
+
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_read_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static int shub_light_write_raw(struct iio_dev *iio_dev,
+				const struct iio_chan_spec *chan, int val,
+				int val2, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_CALIBSCALE:
+		return sprd_shub_calib_write(iio_dev, 0, val);
+
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_write_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static const struct iio_info shub_light_iio_info = {
+	.read_raw = shub_light_read_raw,
+	.write_raw = shub_light_write_raw,
+	.read_avail = sprd_shub_read_avail,
+};
+
+static int sprd_shub_light_probe(struct platform_device *pdev)
+{
+	struct iio_dev *iio_dev;
+	struct sprd_shub_sensor_priv *priv;
+	int ret;
+
+	iio_dev = devm_iio_device_alloc(&pdev->dev, sizeof(*priv));
+	if (!iio_dev)
+		return -ENOMEM;
+
+	priv = iio_priv(iio_dev);
+	priv->sensor = *(struct sprd_shub_sensor **)pdev->dev.platform_data;
+	priv->process = sprd_shub_process_single;
+
+	iio_dev->name = "sprd-shub-light";
+	iio_dev->info = &shub_light_iio_info;
+	iio_dev->channels = shub_light_iio_channels;
+	iio_dev->num_channels = ARRAY_SIZE(shub_light_iio_channels);
+
+	ret = devm_iio_kfifo_buffer_setup(&pdev->dev, iio_dev, &sprd_shub_buffer_ops);
+	if (ret)
+		return ret;
+
+	ret = devm_iio_device_register(&pdev->dev, iio_dev);
+	if (ret)
+		return ret;
+
+	platform_set_drvdata(pdev, iio_dev);
+
+	return 0;
+}
+
+static void sprd_shub_light_remove(struct platform_device *pdev)
+{
+	platform_set_drvdata(pdev, NULL);
+}
+
+static const struct platform_device_id sprd_shub_light_ids[] = {
+	{ .name = "sprd-shub-light" },
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, sprd_shub_light_ids);
+
+static struct platform_driver sprd_shub_light_driver = {
+	.probe = sprd_shub_light_probe,
+	.remove = sprd_shub_light_remove,
+	.driver	= {
+		.name = "sprd_shub_light",
+	},
+	.id_table = sprd_shub_light_ids,
+};
+module_platform_driver(sprd_shub_light_driver);
+
+MODULE_DESCRIPTION("Unisoc sensor hub ambient light sensor driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/iio/magnetometer/Kconfig b/drivers/iio/magnetometer/Kconfig
index fb313e591..0fc77163b 100644
--- a/drivers/iio/magnetometer/Kconfig
+++ b/drivers/iio/magnetometer/Kconfig
@@ -151,6 +151,14 @@ config MMC5633
 	  To compile this driver as a module, choose M here: the module
 	  will be called mmc5633
 
+config IIO_SPRD_SHUB_COMPASS
+	tristate "Unisoc sensor hub compass driver"
+	depends on IIO_SPRD_SHUB
+
+config IIO_SPRD_SHUB_MAGN
+	tristate "Unisoc sensor hub magnetometer driver"
+	depends on IIO_SPRD_SHUB
+
 config IIO_ST_MAGN_3AXIS
 	tristate "STMicroelectronics magnetometers 3-Axis Driver"
 	depends on (I2C || SPI_MASTER) && SYSFS
diff --git a/drivers/iio/magnetometer/Makefile b/drivers/iio/magnetometer/Makefile
index 5bd227f8c..92fd2b865 100644
--- a/drivers/iio/magnetometer/Makefile
+++ b/drivers/iio/magnetometer/Makefile
@@ -17,6 +17,9 @@ obj-$(CONFIG_HID_SENSOR_MAGNETOMETER_3D) += hid-sensor-magn-3d.o
 obj-$(CONFIG_MMC35240)	+= mmc35240.o
 obj-$(CONFIG_MMC5633)	+= mmc5633.o
 
+obj-$(CONFIG_IIO_SPRD_SHUB_COMPASS) += sprd_shub_compass.o
+obj-$(CONFIG_IIO_SPRD_SHUB_MAGN) += sprd_shub_magn.o
+
 obj-$(CONFIG_IIO_ST_MAGN_3AXIS) += st_magn.o
 st_magn-y := st_magn_core.o
 st_magn-$(CONFIG_IIO_BUFFER) += st_magn_buffer.o
diff --git a/drivers/iio/magnetometer/sprd_shub_compass.c b/drivers/iio/magnetometer/sprd_shub_compass.c
new file mode 100644
index 000000000..2c3690184
--- /dev/null
+++ b/drivers/iio/magnetometer/sprd_shub_compass.c
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Unisoc sensor hub IIO compass driver
+ *
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/mod_devicetable.h>
+#include <linux/iio/common/sprd_shub.h>
+#include <linux/iio/kfifo_buf.h>
+#include <linux/platform_device.h>
+
+static const struct iio_chan_spec shub_compass_iio_channels[] = {
+	{
+		.type = IIO_ROT,
+		.modified = true,
+		.channel2 = IIO_MOD_NORTH_MAGN_TILT_COMP,
+		.scan_index = 0,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+	},
+	{
+		.type = IIO_ROT,
+		.modified = true,
+		.channel2 = IIO_MOD_PITCH,
+		.scan_index = 1,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+	},
+	{
+		.type = IIO_ROT,
+		.modified = true,
+		.channel2 = IIO_MOD_ROLL,
+		.scan_index = 2,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+	},
+	{
+		.type = IIO_TIMESTAMP,
+		.channel = -1,
+		.scan_index = 3,
+		.scan_type = {
+			.sign = 'u',
+			.realbits = 64,
+			.storagebits = 64,
+			.endianness = IIO_CPU,
+		},
+	},
+};
+
+static int shub_compass_read_raw(struct iio_dev *iio_dev,
+			      const struct iio_chan_spec *chan, int *val,
+			      int *val2, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_SCALE:
+		*val = 1;
+		*val2 = SPRD_SHUB_FIXED_SCALE;
+		return IIO_VAL_FRACTIONAL;
+
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_read_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static int shub_compass_write_raw(struct iio_dev *iio_dev,
+			       const struct iio_chan_spec *chan, int val,
+			       int val2, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_write_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static const struct iio_info shub_compass_iio_info = {
+	.read_raw = shub_compass_read_raw,
+	.write_raw = shub_compass_write_raw,
+	.read_avail = sprd_shub_read_avail,
+};
+
+static int sprd_shub_compass_probe(struct platform_device *pdev)
+{
+	struct iio_dev *iio_dev;
+	struct sprd_shub_sensor_priv *priv;
+	int ret;
+
+	iio_dev = devm_iio_device_alloc(&pdev->dev, sizeof(*priv));
+	if (!iio_dev)
+		return -ENOMEM;
+
+	priv = iio_priv(iio_dev);
+	priv->sensor = *(struct sprd_shub_sensor **)pdev->dev.platform_data;
+	priv->process = sprd_shub_process_xyz;
+
+	iio_dev->name = "sprd-shub-compass";
+	iio_dev->info = &shub_compass_iio_info;
+	iio_dev->channels = shub_compass_iio_channels;
+	iio_dev->num_channels = ARRAY_SIZE(shub_compass_iio_channels);
+
+	ret = devm_iio_kfifo_buffer_setup(&pdev->dev, iio_dev, &sprd_shub_buffer_ops);
+	if (ret)
+		return ret;
+
+	ret = devm_iio_device_register(&pdev->dev, iio_dev);
+	if (ret)
+		return ret;
+
+	platform_set_drvdata(pdev, iio_dev);
+
+	return 0;
+}
+
+static void sprd_shub_compass_remove(struct platform_device *pdev)
+{
+	platform_set_drvdata(pdev, NULL);
+}
+
+static const struct platform_device_id sprd_shub_compass_ids[] = {
+	{ .name = "sprd-shub-compass" },
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, sprd_shub_compass_ids);
+
+static struct platform_driver sprd_shub_compass_driver = {
+	.probe = sprd_shub_compass_probe,
+	.remove = sprd_shub_compass_remove,
+	.driver	= {
+		.name = "sprd_shub_compass",
+	},
+	.id_table = sprd_shub_compass_ids,
+};
+module_platform_driver(sprd_shub_compass_driver);
+
+MODULE_DESCRIPTION("Unisoc sensor hub compass driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/iio/magnetometer/sprd_shub_magn.c b/drivers/iio/magnetometer/sprd_shub_magn.c
new file mode 100644
index 000000000..813dd9b69
--- /dev/null
+++ b/drivers/iio/magnetometer/sprd_shub_magn.c
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Unisoc sensor hub IIO magnetometer driver
+ *
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/mod_devicetable.h>
+#include <linux/iio/common/sprd_shub.h>
+#include <linux/iio/kfifo_buf.h>
+#include <linux/platform_device.h>
+
+static const struct iio_chan_spec_ext_info shub_magn_ext_info[] = {
+	IIO_MOUNT_MATRIX(IIO_SHARED_BY_ALL, sprd_shub_get_mount_matrix),
+	{ }
+};
+
+static const struct iio_chan_spec shub_magn_iio_channels[] = {
+	{
+		.type = IIO_MAGN,
+		.modified = true,
+		.channel2 = IIO_MOD_X,
+		.scan_index = 0,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+		.ext_info = shub_magn_ext_info,
+	},
+	{
+		.type = IIO_MAGN,
+		.modified = true,
+		.channel2 = IIO_MOD_Y,
+		.scan_index = 1,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+		.ext_info = shub_magn_ext_info,
+	},
+	{
+		.type = IIO_MAGN,
+		.modified = true,
+		.channel2 = IIO_MOD_Z,
+		.scan_index = 2,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE) |
+					    BIT(IIO_CHAN_INFO_SAMP_FREQ),
+		.ext_info = shub_magn_ext_info,
+	},
+	{
+		.type = IIO_TIMESTAMP,
+		.channel = -1,
+		.scan_index = 3,
+		.scan_type = {
+			.sign = 'u',
+			.realbits = 64,
+			.storagebits = 64,
+			.endianness = IIO_CPU,
+		},
+	},
+};
+
+static int shub_magn_read_raw(struct iio_dev *iio_dev,
+			      const struct iio_chan_spec *chan, int *val,
+			      int *val2, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_SCALE:
+		*val = 1;
+		*val2 = SPRD_SHUB_FIXED_SCALE;
+		return IIO_VAL_FRACTIONAL;
+
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_read_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static int shub_magn_write_raw(struct iio_dev *iio_dev,
+			       const struct iio_chan_spec *chan, int val,
+			       int val2, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_write_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static const struct iio_info shub_magn_iio_info = {
+	.read_raw = shub_magn_read_raw,
+	.write_raw = shub_magn_write_raw,
+	.read_avail = sprd_shub_read_avail,
+};
+
+static int sprd_shub_magn_probe(struct platform_device *pdev)
+{
+	struct iio_dev *iio_dev;
+	struct sprd_shub_sensor_priv *priv;
+	int ret;
+
+	iio_dev = devm_iio_device_alloc(&pdev->dev, sizeof(*priv));
+	if (!iio_dev)
+		return -ENOMEM;
+
+	priv = iio_priv(iio_dev);
+	priv->sensor = *(struct sprd_shub_sensor **)pdev->dev.platform_data;
+	priv->process = sprd_shub_process_xyz;
+
+	iio_dev->name = "sprd-shub-magn";
+	iio_dev->info = &shub_magn_iio_info;
+	iio_dev->channels = shub_magn_iio_channels;
+	iio_dev->num_channels = ARRAY_SIZE(shub_magn_iio_channels);
+
+	ret = devm_iio_kfifo_buffer_setup(&pdev->dev, iio_dev, &sprd_shub_buffer_ops);
+	if (ret)
+		return ret;
+
+	ret = devm_iio_device_register(&pdev->dev, iio_dev);
+	if (ret)
+		return ret;
+
+	platform_set_drvdata(pdev, iio_dev);
+
+	return 0;
+}
+
+static void sprd_shub_magn_remove(struct platform_device *pdev)
+{
+	platform_set_drvdata(pdev, NULL);
+}
+
+static const struct platform_device_id sprd_shub_magn_ids[] = {
+	{ .name = "sprd-shub-magn" },
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, sprd_shub_magn_ids);
+
+static struct platform_driver sprd_shub_magn_driver = {
+	.probe = sprd_shub_magn_probe,
+	.remove = sprd_shub_magn_remove,
+	.driver	= {
+		.name = "sprd_shub_magn",
+	},
+	.id_table = sprd_shub_magn_ids,
+};
+module_platform_driver(sprd_shub_magn_driver);
+
+MODULE_DESCRIPTION("Unisoc sensor hub magnetometer driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/iio/proximity/Kconfig b/drivers/iio/proximity/Kconfig
index bb77fad2a..c6340d308 100644
--- a/drivers/iio/proximity/Kconfig
+++ b/drivers/iio/proximity/Kconfig
@@ -55,6 +55,10 @@ config HX9023S
 	  To compile this driver as a module, choose M here: the
 	  module will be called hx9023s.
 
+config IIO_SPRD_SHUB_PROXIMITY
+	tristate "Unisoc sensor hub proximity sensor driver"
+	depends on IIO_SPRD_SHUB
+
 config IRSD200
 	tristate "Murata IRS-D200 PIR sensor"
 	select IIO_BUFFER
diff --git a/drivers/iio/proximity/Makefile b/drivers/iio/proximity/Makefile
index 4352833dd..d720878f7 100644
--- a/drivers/iio/proximity/Makefile
+++ b/drivers/iio/proximity/Makefile
@@ -8,6 +8,7 @@ obj-$(CONFIG_AS3935)		+= as3935.o
 obj-$(CONFIG_CROS_EC_MKBP_PROXIMITY) += cros_ec_mkbp_proximity.o
 obj-$(CONFIG_D3323AA)		+= d3323aa.o
 obj-$(CONFIG_HX9023S)		+= hx9023s.o
+obj-$(CONFIG_IIO_SPRD_SHUB_PROXIMITY) += sprd_shub_proximity.o
 obj-$(CONFIG_IRSD200)		+= irsd200.o
 obj-$(CONFIG_ISL29501)		+= isl29501.o
 obj-$(CONFIG_LIDAR_LITE_V2)	+= pulsedlight-lidar-lite-v2.o
diff --git a/drivers/iio/proximity/sprd_shub_proximity.c b/drivers/iio/proximity/sprd_shub_proximity.c
new file mode 100644
index 000000000..cd455bc44
--- /dev/null
+++ b/drivers/iio/proximity/sprd_shub_proximity.c
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Unisoc sensor hub IIO proximity sensor driver
+ *
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/mod_devicetable.h>
+#include <linux/iio/buffer.h>
+#include <linux/iio/common/sprd_shub.h>
+#include <linux/iio/kfifo_buf.h>
+#include <linux/platform_device.h>
+#include <linux/unaligned.h>
+
+static const struct iio_chan_spec shub_proximity_iio_channels[] = {
+	{
+		.type = IIO_PROXIMITY,
+		.scan_index = 0,
+		.scan_type = {
+			.sign = 's',
+			.realbits = 32,
+			.storagebits = 32,
+			.endianness = IIO_CPU,
+		},
+		.info_mask_separate = BIT(IIO_CHAN_INFO_SCALE) |
+				      BIT(IIO_CHAN_INFO_SAMP_FREQ),
+	},
+	{
+		.type = IIO_TIMESTAMP,
+		.channel = -1,
+		.scan_index = 1,
+		.scan_type = {
+			.sign = 'u',
+			.realbits = 64,
+			.storagebits = 64,
+			.endianness = IIO_CPU,
+		},
+	},
+};
+
+static int shub_proximity_read_raw(struct iio_dev *iio_dev,
+				   const struct iio_chan_spec *chan, int *val,
+				   int *val2, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_SCALE:
+		*val = 1;
+		return IIO_VAL_INT;
+
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_read_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static int shub_proximity_write_raw(struct iio_dev *iio_dev,
+				    const struct iio_chan_spec *chan, int val,
+				    int val2, long mask)
+{
+	switch (mask) {
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return sprd_shub_write_freq(iio_dev, val, val2);
+	}
+
+	return -EINVAL;
+}
+
+static const struct iio_info shub_proximity_iio_info = {
+	.read_raw = shub_proximity_read_raw,
+	.write_raw = shub_proximity_write_raw,
+	.read_avail = sprd_shub_read_avail,
+};
+
+static void shub_proximity_process(struct iio_dev *iio_dev, void *data, u64 timestamp)
+{
+	struct sprd_shub_sensor_priv *priv = iio_priv(iio_dev);
+	s32 buf[4];
+
+	if (test_bit(0, &priv->scan_mask))
+		buf[0] = get_unaligned_le32(data) ? 0 : 1;
+
+	iio_push_to_buffers_with_timestamp(iio_dev, buf, timestamp);
+}
+
+static int sprd_shub_proximity_probe(struct platform_device *pdev)
+{
+	struct iio_dev *iio_dev;
+	struct sprd_shub_sensor_priv *priv;
+	int ret;
+
+	iio_dev = devm_iio_device_alloc(&pdev->dev, sizeof(*priv));
+	if (!iio_dev)
+		return -ENOMEM;
+
+	priv = iio_priv(iio_dev);
+	priv->sensor = *(struct sprd_shub_sensor **)pdev->dev.platform_data;
+	priv->process = shub_proximity_process;
+
+	iio_dev->name = "sprd-shub-proximity";
+	iio_dev->info = &shub_proximity_iio_info;
+	iio_dev->channels = shub_proximity_iio_channels;
+	iio_dev->num_channels = ARRAY_SIZE(shub_proximity_iio_channels);
+
+	ret = devm_iio_kfifo_buffer_setup(&pdev->dev, iio_dev, &sprd_shub_buffer_ops);
+	if (ret)
+		return ret;
+
+	ret = devm_iio_device_register(&pdev->dev, iio_dev);
+	if (ret)
+		return ret;
+
+	platform_set_drvdata(pdev, iio_dev);
+
+	return 0;
+}
+
+static void sprd_shub_proximity_remove(struct platform_device *pdev)
+{
+	platform_set_drvdata(pdev, NULL);
+}
+
+static const struct platform_device_id sprd_shub_proximity_ids[] = {
+	{ .name = "sprd-shub-proximity" },
+	{ }
+};
+MODULE_DEVICE_TABLE(platform, sprd_shub_proximity_ids);
+
+static struct platform_driver sprd_shub_proximity_driver = {
+	.probe = sprd_shub_proximity_probe,
+	.remove = sprd_shub_proximity_remove,
+	.driver	= {
+		.name = "sprd_shub_proximity",
+	},
+	.id_table = sprd_shub_proximity_ids,
+};
+module_platform_driver(sprd_shub_proximity_driver);
+
+MODULE_DESCRIPTION("Unisoc sensor hub proximity sensor driver");
+MODULE_LICENSE("GPL");
diff --git a/include/linux/iio/common/sprd_shub.h b/include/linux/iio/common/sprd_shub.h
new file mode 100644
index 000000000..8fc7a786c
--- /dev/null
+++ b/include/linux/iio/common/sprd_shub.h
@@ -0,0 +1,38 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#ifndef __SPRD_SHUB_H__
+#define __SPRD_SHUB_H__
+
+#include <linux/iio/iio.h>
+
+struct sprd_shub_sensor_priv {
+	struct sprd_shub_sensor *sensor;
+	void (*process)(struct iio_dev *iio_dev, void *data, u64 timestamp);
+
+	unsigned long scan_mask;
+	int freq_idx;
+	s32 calib_data[32];
+};
+
+extern const struct iio_buffer_setup_ops sprd_shub_buffer_ops;
+
+const struct iio_mount_matrix *
+sprd_shub_get_mount_matrix(const struct iio_dev *iio_dev,
+			   const struct iio_chan_spec *chan);
+
+int sprd_shub_calib_read(struct iio_dev *iio_dev, int idx, int *val);
+int sprd_shub_calib_write(struct iio_dev *iio_dev, int idx, int val);
+
+int sprd_shub_read_freq(struct iio_dev *iio_dev, int *val, int *val2);
+int sprd_shub_write_freq(struct iio_dev *iio_dev, int val, int val2);
+int sprd_shub_read_avail(struct iio_dev *iio_dev,
+			 struct iio_chan_spec const *chan,
+			 const int **vals, int *type,
+			 int *length, long mask);
+
+#define SPRD_SHUB_FIXED_SCALE	256
+
+void sprd_shub_process_xyz(struct iio_dev *iio_dev, void *data, u64 timestamp);
+void sprd_shub_process_single(struct iio_dev *iio_dev, void *data, u64 timestamp);
+
+#endif /* __SPRD_SHUB_H__ */
